src/box/lua/net_box.lua

```diff
@@ -390,6 +391,24 @@ local function new_sm(uri, opts)
                     })
                 end
             end
+            if not opts.fetch_schema and features.space_and_index_names then
+                remote.space = setmetatable({_remote = remote}, {
+                    __index = function(self, key)
+                        local space = {id = key, name = key,
+                                       _remote = self._remote,
+                                       _format_cdata =
+                                            box.internal.new_tuple_format()}
+                        space.index = setmetatable({space = space,
+                                                    _remote = remote}, {
+                            __index = function(self, key)
+                                return setmetatable({id = key, name = key,
+                                                     space = self.space,
+                                                     _remote = self._remote},
+                                                    self._remote._index_mt)
+                            end})
+                        return setmetatable(space, self._remote._space_mt)
```

Comment:
AFAIR we agreed to use weak tables to avoid creation of a space and index objects each time they are accessed.

Reply:
I don't understand how you want to use weak tables here. I thought we wanted to make the `c.space` and `c.space.s.index` tables have weak values, so that the connection objects have weak references, but it doesn't help, since their metamethods are closures that capture connection objects.

Reply:
Looks like the problem was caused by the typo I made that you mention two comments bellow (I used `remote` instead of `self._remote` and it created an upvalue). Now setting `__mode = 'v'` for `c.space` and `c.space.s.index` works as expected (except for the latter, the space field will always get garbage collected, which I don't think we want to happen).

> to avoid creation of a space and index objects each time they are accessed.

I still don't understand how you want to use weak tables here with this goal, could you please elaborate a bit or redirect me to some example?

Reply:
Discussed with @locker and @alyapunov: the space and object tables shall be register tables (i.e., cache the values they are indexed with) with strong references — all necessary context stored as upvalues.